### ISSUE TITLE
Are you planning to implement LiveQuery in dotnet sdk? 

### ISSUE THREAD

**Issue by Mahdi7s** [parse-community/Parse-SDK-dotNET#199]
we are developing a real-time game using unity3d . please let us know that if you are going to implement the LiveQuery feature in your sdk, so we start using it to develop our game.

**Comment by richardjrossiii** [comment]
Eventually, yes. We do have plans to support LiveQuery on all of our targets that we currently support for .NET.

The situation is a bit more complicated than other platforms, as we have such a diversity in versions of .NET and APIs supported by our .NET platforms compared to everywhere else, and as such we do not currently have an ETA. 

I'll keep this issue open until we have a formal release for it, or when we might have an ETA.

To others reading this issue: To help us gauge interest and priority amongst other issues, if you're interested in this, please comment on this issue.

**Comment by IkeTheDestroyer** [comment]
Definitely interested in this for Unity iOS/Android.

**Comment by thewizster** [comment]
:+1: :+1:

**Comment by arturogutierrez** [comment]
💯

**Comment by hl4hck** [comment]
+1

**Comment by Apoll0** [comment]
+1 for Unity

**Comment by hl4hck** [comment]
I am developing a mobile game using Unity3D.
Is it possible to use LiveQuery with REST API (HTTP.Request)?

**Comment by richardjrossiii** [comment]
@hl4hck Absolutely! The protocol specification for LiveQuery can be found here: https://github.com/ParsePlatform/parse-server/wiki/Parse-LiveQuery-Protocol-Specification.

You can also, of course, take a look at the open source JS and iOS clients and use those as a basis for any other clients you might make.

**Comment by hl4hck** [comment]
@richardjrossiii Thank you!

**Comment by couchworx** [comment]
+1

**Comment by kenglou** [comment]
+1

**Comment by chris3000** [comment]
+1, although the API seems pretty simple. I'll try to implement in Unity using BestHTTP until it's stable in the Parse SDK.

**Comment by abagshaw** [comment]
+1

**Comment by johnnydimas** [comment]
yes please!

**Comment by FlorianSchuler** [comment]
+1

**Comment by tofutim** [comment]
Has anyone successfully gotten LiveQuery to work from .NET?

**Comment by farzadi** [comment]
+1

**Comment by miracle7** [comment]
@chris3000 Did you have any luck getting this working with Unity?

Looking into trying it myself but if there is anything interesting you have to share could save some tears haha

**Comment by chris3000** [comment]
No, I haven't tried yet. We're using a polling strategy for now.

**Comment by plyoung** [comment]
+1 Unity.

**Comment by vladrad** [comment]
I would be so down to work on this! Im just not sure where to start.

**Comment by Apoll0** [comment]
+1 for unity

**Comment by ndastur** [comment]
+1 Would love to see this for xamarin forms apps

**Comment by eclipse1985** [comment]
+1

**Comment by muhammad-naderi** [comment]
+1

**Comment by ra7bi** [comment]
+1 any updates ?

**Comment by montymxb** [comment]
Just to let everyone know there is currently no built-in live query implementation available for the .net sdk. Only some clients have implementations for live query interfacing, this not being one of them. 

That being said it's not unlikely that we will bring this support into the this sdk at a future date. It would be super to have this in! However this sdk needs to be brought up to spec first as there are quite a few outstanding issues that need to be dealt with before we can get to live queries.

**Comment by ra7bi** [comment]
i wish more contributors comes to improve Parse & Parse SDK , Really nice and great server , just it need someone who have good experience to improve it

**Comment by montymxb** [comment]
@ra7bi we're open to contributions! Most of the sdks are more actively maintained, this one just needs some tlc and it should be just fine 👍

**Comment by ra7bi** [comment]
happy to know that , im hosting Parse server for my mobile app and i really like it , i think if Real-Time feature comes to Parse server for all SDK ,  it would be the best . 

Thanks for all Parse contributions so exciting for any Parse updates

**Comment by stephenhauck** [comment]
I need examples of how to use live query in a Xamarin Forms app....

**Comment by ndastur** [comment]
As above. Not implemented

**Comment by stephenhauck** [comment]
OK, thanks for confirming... I will select another backend for my needs that supports Xamarin Forms.

**Comment by flovilmart** [comment]
@stephenhauck, you can still use parse-server without live query, we have 3 reference implementations in JS, Android and Swift porting to Xamarin/dotNet should be trivial.

**Comment by ndastur** [comment]
@flovilmart interesting comment, as if it was trivial I would assume it wouldn't be an outstanding ticket for 2 years ;)
I think the honest answer is the .NET parse SDK is a bit neglected.

**Comment by flovilmart** [comment]
@ndastur it’s neglected because unlike the other SDK’s no one has demonstrated an interest into contributing with it. Open source is not just about a handful contributors shipping features you need. We ported to Android in a few days, when our company needed it.

I’m not a dotNet developer, and my main focus is on server, JS, Swift, objective-c, and i believe we still move forward at a great pace. Like any open source project, we’ll be thrilled to review your pull request and fast track a release. Also, in Java and iOS, live Query is not part of the SDK but distributed as a separate package, Which also give you the freedom to build your own implementation without the burden of open sourcing it.

**Comment by flovilmart** [comment]
@ndastur most of the PR's were awaiting additional changes, from @richardjrossiii. I'm not sure why, I'm not a .Net developer but you seem to be well versed into this technology. Can you review those for me please? If there's any smell / issue, feel free to ping me.

**Comment by stephenhauck** [comment]
I appreciate this is an open source effort but Back4app sent me here, they are charging for a Parse back end and I kind of expected that they were supporting the project / product.

I am looking to pay for Back4app as a service .. that does not put the onus on me to make the drivers that support THEIR product.... I will simply find a vendor that wants my money and supports my platform.

**Comment by flovilmart** [comment]
@stephenhauck I understand your frustration, and the open source community is not the place to project it. You are right, if a vendor doesn't fulfill exactly your needs, you can probably leave them. That's not my problem. Just so you know, parse-community is completely independently managed from the vendors, we follow our own roadmap and goals. 

> that does not put the onus on me to make the drivers that support THEIR product
You are 100% right, but don't blame the community for it.

Perhaps the style of parse-server, being an open source project, packaged by vendors as a paying service is not suiting your needs and development requirement. 

@baywet is it something you could give us a hand on?

**Comment by stephenhauck** [comment]
@flovilmart Not projecting frustration .. simply stating the facts.

You guys are doing awesome work ... I appreciate your efforts even if I / my applications do not benefit from them at this time ...

**Comment by baywet** [comment]
I might, I'm trying to get an understanding of what needs to be implemented and the amount of work we're talking about.
We're basically looking at implementing support for that, but in dotnet right? http://docs.parseplatform.org/parse-server/guide/#client-setup
And all the events and such are pushed by the server using a web socket?

**Comment by flovilmart** [comment]
Yep, we're looking at that spec (https://github.com/parse-community/parse-server/wiki/Parse-LiveQuery-Protocol-Specification) 

Reference implementations in [JS](https://github.com/parse-community/Parse-SDK-JS/blob/master/src/LiveQueryClient.js#L125), [Swift](https://github.com/parse-community/ParseLiveQuery-iOS-OSX/blob/master/Sources/ParseLiveQuery/Client.swift#L20) and [Android/Java](https://github.com/parse-community/ParseLiveQuery-Android/blob/master/ParseLiveQuery/src/main/java/com/parse/ParseLiveQueryClientImpl.java#L24) 

Thanks man :)

**Comment by baywet** [comment]
I didn't say I'll do it yet but at least any senior .net Dev will have all the information here.
And then how to test the code? Is there like an open server/service somewhere? Or just follow that and start sending data ? http://docs.parseplatform.org/parse-server/guide/

**Comment by flovilmart** [comment]
We don't have a conformance spec testing, but firing up the server and going for e2e tests should be good enough for now.

Just thank you for taking the time to respond, and layout a plan in here!

**Comment by montymxb** [comment]
I'm planning on going over PRs as soon as I can make enough time. It is about time that we bring this into the fold with the other sdks, especially since there is still a community interest in it.

@baywet if you are interested in working on this let me know, although I'm not _quite_ familiar with .NET development I work on the server and some other sdks. Let me know if I can provide some insight as to how this is laid out and how it operates.

**Comment by baywet** [comment]
@montymxb my biggest problem right now is time, and I don't want to over commit. As far as I can see there's somewhere between 5 to 10 hours of effort on this one, and I have to ramp up on parse. I don't pretend to be a Rockstar developer but I

**Comment by flovilmart** [comment]
@baywet, no stress, @montymxb, i met @baywet at a Microsoft MVP event last week, in Montreal. 

Perhaps we know a junior dev who’d like to grind his teeths on this one

**Comment by montymxb** [comment]
@baywet that's perfectly fine. If you ever find you have a little bit of time and happen to be strolling by parse land, feel free to drop in and give us some notes in regards to .NET stuff. We're always open ears! But definitely don't stretch yourself too thin if you have existing priorities that are demanding your attention. It may be that you won't have the time to drop by, but that's better than burning yourself out 😛 .

I don't know too many .NET devs unfortunately :/. There was one guy, but he works full time at microsoft now and I think he's been pretty busy lately.

**Comment by TheFanatr** [comment]
I have been working on this for the past while, without noticing that people started posting to this issue again. I fit the description that @flovilmart mentioned in his previous post, and I also happen to be on a break from school right now, so I'll take on the job officially if you guys want. I'm actually close to finishing a basic implementation, so I'll post when there's something new.

**Comment by baywet** [comment]
@TheFanatr that's great. Don't hesitate to mention me on the PR once you submit it. I'll help reviewing the code :)

**Comment by drinvojka** [comment]
any updates about .net support ?

**Comment by TheFanatr** [comment]
@drinvojka If you mean .NET Standard, the master branch of this repository already supports it; you just have to build it from source after cloning or downloading and reference the output assembly, or you can go to the AppVeyor build and download the NuGet package from the artifacts section and reference that instead. If you would like, I can provide a direct link to it, but I would recommend just building from source because it's less annoying than setting up a local NuGet package source/repository just so that you can reference one package. As for stability, the project should work as before, but it is still considered pre-alpha 2.0.0, so tread carefully.

**Comment by TheFanatr** [comment]
It also occurred to me that if you are referring to LiveQuery specifically, I was working on it at some point, but did not have enough time to finish it. I may at some point, but it should be noted that pretty much anyone can do it because the .NET/CLI API is almost analogous to the Java/JVM API, so it shouldn't require too much work.

**Comment by JonMcPherson** [comment]
Is this being worked on at all? any plans to release it into this library?  I'm thinking about implementing live query support myself if it doesn't yet exist as open source.  I'll probably base it off of the Android/Java implementation.

**Comment by flovilmart** [comment]
@JonMcPherson feel free to get started, I believe the architecture of the SDK is settled now and @montymxb could get you started!

**Comment by JonMcPherson** [comment]
@flovilmart Just to update, I have ported the ParseLiveQuery-Android project to .NET Standard 2.0 using the [websocket-sharp](https://github.com/sta/websocket-sharp) library complete with unit tests.
I am currently working on integrating this with a Unity 2018.2 project, and will update here once I get that working.
https://github.com/JonMcPherson/parse-live-query-dotnet

**Comment by flovilmart** [comment]
That’s awesome! Wonderful!

**Comment by bothwellw** [comment]
Hi All, is LiveQuery now available on the .NET SDK by anychance ?

**Comment by joechihe** [comment]
I think this topic is not complicated to implement. I'm using Back4app as Parse cloud provider.
In Back4app enabling LiveQuery is extremely easy.
Then using websocket-sharp library, in less than 1 hour, you can connect to LiveQuery and receive real time notifications. It is very simple. I already was able to do it and I will share an example soon.
@JonMcPherson I checked your code: https://github.com/JonMcPherson/parse-live-query-dotnet. Seems to be very promising, however I think the implementation is little bit over complicated for most scenarios. Still is a good to check it to understand the connection against Parse Server service.

**Comment by TheFanatr** [comment]
@joechihe This needs to be done; if you believe you can, please make a pull request adding this feature. Otherwise, I will eventually get to it, but it will take a while.

**Comment by joechihe** [comment]
> @joechihe This needs to be done; if you believe you can, please make a pull request adding this feature. Otherwise, I will eventually get to it, but it will take a while.

@TheFanatr Oh, I didn't mean to wait until your can implement. Actually I wanted to share that everyone can implement this, and not to wait for you guys to add the feature. You're doing great job! And Parse Server is still a great solution for a backend system

**Comment by TheFanatr** [comment]
We'll of course still try our best to add it. 😃

**Comment by JonMcPherson** [comment]
@joechihe Yea I mean porting the Java implementation was easy enough, and it works great for me.  That's also the reason the implementation seems a bit verbose and over-complicated. Had I implemented a client myself, I probably would have trimmed it down a bit.

**Comment by joechihe** [comment]
> @joechihe Yea I mean porting the Java implementation was easy enough, and it works great for me. That's also the reason the implementation seems a bit verbose and over-complicated. Had I implemented a client myself, I probably would have trimmed it down a bit.

Cool, I will try to create a simplified implementation with the most common scenarios. Still I don't know if it is easier create it as separated library, or create it as functional feature inside Parse SDK library. I have limited time, so I need to go for easiest option for me. 
@TheFanatr If I can include it as part of this library, under what namespace or folder is better to add it?

**Comment by JanusMirith** [comment]
I have LiveQuiry working in unity 2020.1 on the new 2.0.0 code base.
I just grabbed Jon's code and fixed all the breaking changes.
I plan to move forward with cleaning things up and moving things more to events rather then callbacks.

If this is interesting to anyone I am happy to share, just not quite sure the best way to do so.

**Comment by Evilc06** [comment]
> I have LiveQuiry working in unity 2020.1 on the new 2.0.0 code base.
> I just grabbed Jon's code and fixed all the breaking changes.
> I plan to move forward with cleaning things up and moving things more to events rather then callbacks.
> 
> If this is interesting to anyone I am happy to share, just not quite sure the best way to do so.

If you do manage to make it work please do let me know, I would like to test this for Unity and develop something with it :)

**Comment by JanusMirith** [comment]
> If you do manage to make it work please do let me know, I would like to test this for Unity and develop something with it :)

It is working now, I am just writing up some quality of life improvements for unity (automatically calling on main thread)
I am also doing some automatic local caching and whatnot to make it kinda run like firebase, probably not to everyone's tastes.
If no-one is in a huge rush I'll take a week to test and comment everything then share

**Comment by Evilc06** [comment]
> > If you do manage to make it work please do let me know, I would like to test this for Unity and develop something with it :)
> 
> It is working now, I am just writing up some quality of life improvements for unity (automatically calling on main thread)
> I am also doing some automatic local caching and whatnot to make it kinda run like firebase, probably not to everyone's tastes.
> If no-one is in a huge rush I'll take a week to test and comment everything then share

Not a problem, take your time to test and possibly comment that would be truly appreciated. Also, would you be able to provide a test on how to make it work?

**Comment by JanusMirith** [comment]
Ok I have it all running with a test scene and a example script (just fill in your own keys and URLs)
It is currently designed to sit in the asset folder as a submodule, perhaps in future it might be a package.

Effectively it is just a tweak of what @JonMcPherson did with some quality of life stuff for unity, I would love to maybe work out a way to properly contribute to that or to this. But at the moment just pull this down and let me know what you think.
[https://github.com/JanusMirith/ParseRealtimeQuery](url)

I have it working on Windows Standalone, iOS, Android (including Quest (woooo))
It is NOT working on UWP (hololens, boooooo) due to a possible Parse bug (can not access C:// on initialize (something to do with the Cache Controller))

If all that you want is a Parse 2.0.0 compatible liveQuery implementation you can ignore the stuff around RealTimeQuery, that whole thing is a way i wanted to simplify the process of a liveQuery by combining the initial query and the live query into one Real Time entity that you can just make subscribe to. I think it is fabulous but it might be hot garbage for all I know.

**Comment by Evilc06** [comment]
Awesome thanks, will test it out when I can and let you know how it goes!

**Comment by jeansebastienZ** [comment]
> Ok I have it all running with a test scene and a example script (just fill in your own keys and URLs)
> It is currently designed to sit in the asset folder as a submodule, perhaps in future it might be a package.
> 
> Effectively it is just a tweak of what @JonMcPherson did with some quality of life stuff for unity, I would love to maybe work out a way to properly contribute to that or to this. But at the moment just pull this down and let me know what you think.
> [https://github.com/JanusMirith/ParseRealtimeQuery](url)
> 
> I have it working on Windows Standalone, iOS, Android (including Quest (woooo))
> It is NOT working on UWP (hololens, boooooo) due to a possible Parse bug (can not access C:// on initialize (something to do with the Cache Controller))
> 
> If all that you want is a Parse 2.0.0 compatible liveQuery implementation you can ignore the stuff around RealTimeQuery, that whole thing is a way i wanted to simplify the process of a liveQuery by combining the initial query and the live query into one Real Time entity that you can just make subscribe to. I think it is fabulous but it might be hot garbage for all I know.



> Ok I have it all running with a test scene and a example script (just fill in your own keys and URLs)
> It is currently designed to sit in the asset folder as a submodule, perhaps in future it might be a package.
> 
> Effectively it is just a tweak of what @JonMcPherson did with some quality of life stuff for unity, I would love to maybe work out a way to properly contribute to that or to this. But at the moment just pull this down and let me know what you think.
> [https://github.com/JanusMirith/ParseRealtimeQuery](url)
> 
> I have it working on Windows Standalone, iOS, Android (including Quest (woooo))
> It is NOT working on UWP (hololens, boooooo) due to a possible Parse bug (can not access C:// on initialize (something to do with the Cache Controller))
> 
> If all that you want is a Parse 2.0.0 compatible liveQuery implementation you can ignore the stuff around RealTimeQuery, that whole thing is a way i wanted to simplify the process of a liveQuery by combining the initial query and the live query into one Real Time entity that you can just make subscribe to. I think it is fabulous but it might be hot garbage for all I know.

Hello, i can't open the scene with unity ( i'm using mac , don't know if it imply something ) . By the way , could you explain , how to import it on an existing project ?

Thanks you ,

**Comment by jeansebastienZ** [comment]
Did someone made JanusMirith job work ? If yes , could you explain how you did import the project , where to put the files ?

**Comment by zhaosizheng** [comment]
I tried to build JonMcPherson's code on  https://github.com/JonMcPherson/parse-live-query-dotnet with visual studio 2019, 
but failed because cannot find Parse.Common.Internal and Parse.Core.Internal, anyone have any idea what's going wrong?

**Comment by zhaosizheng** [comment]
I try to use parse livequery dotnet on my winform project so my winform can get real time data from mongo database, but I just cannot solve the issues on  https://github.com/JonMcPherson/parse-live-query-dotnet which failed to build, anybody can help me out will be very appreciated, my email is zhaosizheng@yahoo.com

**Comment by jorgeluis11** [comment]
+1 Unity - C#

**Comment by hamednasery** [comment]
+1 c# with Unity - Thanks guys

**Comment by YBTopaz8** [comment]
I did an Initial Port of Live Queries from [JonMcPherson](https://github.com/JonMcPherson/parse-live-query-dotnet) in .NET that works on .net 9 and MAUI. 
I tested in in Windows and Android but don't think other platforms will be an issue..

https://github.com/YBTopaz8/parse-live-query-dotnet

I even did a video on how to use, if that can help!
https://youtu.be/V-cUjq7Js84

**Comment by patkozlowski** [comment]
Thanks @YBTopaz8, great work.  Will be testing this soon.  I have a need for iOS devices (for Maui), but seeing this, I think it should work with that as well.

**Comment by YBTopaz8** [comment]
Hi @patkozlowski appreciate reply!
I can AND will test on MacOS likely by end of weekend but I can't test on an iPhone sadly since I don't have a dev account there.
I can always explain my whole workflow when updating/upgrading the SDK/live queries any time needed.

Despite being confident that it works well (after bunch of testing online, security verifications for mutex/atomicity in the new implementation of .NET 9), anyone is free to reach out in any way.

As expected, I did a sort of "soft rework" of the original parse SDK too, just by switching previous ways (UnWrapp(),OnSuccess etc) to the new-streamlined tasks way (always hated them lol!). 

Been doing this port for months on end and am happy to share now finally (especially since I got it running for clients for a while and no complaints!) 
I forgot to add my contact in previous reply if someone wants to reach out - sorry! it's 8brunel@gmail.com

Glad to help @patkozlowski  @mtrezza  !


(tagging some of others for them to see as well since they were waiting for such a long time lol! @hamednasery  @zhaosizheng  )

**Comment by YBTopaz8** [comment]
Hi everyone,
I’ve submitted a PR that addresses this issue by upgrading the SDK to .NET 9 and introducing Live Queries support for .NET platforms, including .NET MAUI. 
I look forward to your feedback and collaboration!

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [4.0.0](https://github.com/parse-community/Parse-SDK-dotNET/releases/tag/4.0.0)

**Comment by mtrezza** [comment]
This feature is still not implemented; reopening.

**Comment by YBTopaz8** [comment]
All things going well, .NET should be beginning 2025 with Parse LQ support with RX.Net built-in. 

We just needed to do a major update
